### PR TITLE
Transaction support + MQTT connection stabilzation

### PR DESCRIPTION
**Pull request by lt-holman**
This commit contains a few fixes in 2 main areas.

1)  This adds the api `iot_transaction_status` which allows the user to be able to see the status of a publish of a telemetry, attribute, or event...  when the cloud has handled that piece of information
2) This fixes connection issues that are experienced after being alive for long periods of time, but fixing the underlying mqtt sub-system.

### CODE REVIEW

File: src/api/iot_base.c [hunk 1]
```diff
@@ -549,6 +549,9 @@ iot_status_t iot_base_device_id_set(
 				os_file_close( fd );
 				if( device_id_len >= IOT_ID_MAX_LEN )
 					device_id_len = IOT_ID_MAX_LEN;
+				if ( device_id_len > 0u &&
+					device_id[device_id_len] == '\n' )
```
Comment (element-82) [inline src/api/iot_base.c]: why put a line break there?

File: src/api/iot_mqtt.c [hunk 2]
```diff
@@ -299,203 +312,398 @@ iot_mqtt_t* iot_mqtt_connect(
...
-						result->mosq,
-						ssl_conf->insecure );
-				}
+			if ( port == IOT_MQTT_PORT_WSS ||
+				( opts->websocket_path && opts->ssl_conf ) )
+			{
+				uri_proto = "wss";
+				if ( opts->websocket_path &&
+					*opts->websocket_path != '\0' )
+					ws_path = opts->websocket_path;
```
Comment (element-82) [inline src/api/iot_mqtt.c]: what is the websocket path set to?

File: src/api/iot_mqtt.c [hunk 3]
```diff
@@ -504,54 +712,41 @@ iot_status_t iot_mqtt_disconnect(
...
 		mosquitto_destroy( mqtt->mosq );
 		mqtt->mosq = NULL;
 #else /* ifdef IOT_MQTT_MOSQUITTO */
-		mqtt->connected = IOT_FALSE;
 #ifdef IOT_THREAD_SUPPORT
 		{
 			MQTTAsync_disconnectOptions opts =
 				MQTTAsync_disconnectOptions_initializer;
-			opts.timeout = IOT_MQTT_KEEP_ALIVE;
+			opts.timeout = 60; /* wait for up to 60 seconds */
```
Comment (element-82) [inline src/api/iot_mqtt.c]: define?

File: src/api/iot_mqtt.c [hunk 4]
```diff
@@ -504,54 +712,41 @@ iot_status_t iot_mqtt_disconnect(
...
 				== MQTTASYNC_SUCCESS )
 				result = IOT_STATUS_SUCCESS;
 			MQTTAsync_destroy( &mqtt->client );
 		}
-		os_thread_condition_destroy( &mqtt->notification_signal );
-		os_thread_mutex_destroy( &mqtt->notification_mutex );
 #else /* ifdef IOT_THREAD_SUPPORT */
-		if ( MQTTClient_disconnect( mqtt->client, IOT_MQTT_KEEP_ALIVE )
+		/* wait for up to 60 seconds */
+		if ( MQTTClient_disconnect( mqtt->client, 60 )
```
Comment (element-82) [inline src/api/iot_mqtt.c]: define?

File: src/api/iot_base.c (hunk 1 above)
Comment (lt-holman) [reply src/api/iot_base.c]: I'm not, it's checking for a '\n'.  It removes it if so.  In case the line ends with a  '\n'

**Pull request merged**